[WNMGDS-637] Update release process guide (#823)
CMSgov/design-system: guides/RELEASE-PROCESS.md

      **Note**: Your NPM account must access to the CMS group and provide a valid access token. To add a token, edit your `~/.npmrc` file so the contents are `//registry.npmjs.org/:_authToken={token}`

>    1. Run the prerelease script to automatically prepare the NPM package for publishing.

Reviewer: ```suggestion
   1. Run the prerelease script to prepare the repo for publishing.
```

      ```
      yarn pre-release
      ```

      This script will install dependencies, build design system assets, and run the linter and tests. If everything passes, `lerna` will prompt you for the new version number, which should follow the [SemVer release format](#versioning). After the versions are updated, the script will create a tagged release commit (prefaced with `core-`), a branch (prefaced with `release-`) containing the release commit, and a zip for the Github release notes.

>    1. After the prepublish script is completed, visually check that release zip created from the release script (i.e. `cmsgov-design-system-2.0.0.tgz`) contains the folders and files

Reviewer 2: Maybe we can be more explicit here and say what folders and files should be in the zip?

>    1. After the prerelease script is completed, unpack the release zips created from the release script (i.e. `cmsgov-design-system-2.0.0.tgz`) and check the relevant folders and files are present. Confirm that the cmsgov-design-system release zip contains the `dist` folder.

Reviewer: This is great!

      **Note**: Your NPM account must access to the CMS group and provide a valid access token. To add a token, edit your `~/.npmrc` file so the contents are `//registry.npmjs.org/:_authToken={token}`

   1. Run the prepublish script to automatically prepare the NPM package for publishing.

      ```
      yarn pre-release
      ```

      This script will install dependencies, build design system assets, and run the linter and tests. If everything passes, `lerna` will prompt you for the new version number, which should follow the [SemVer release format](#versioning). After the versions are updated, the script will create a tagged release commit (prefaced with `core-`), a branch (prefaced with `release-`) containing the release commit, and a zip for the Github release notes.

>    1. After the prepublish script is completed, visually check that release zip created from the release script (i.e. `cmsgov-design-system-2.0.0.tgz`) contains the relevant folders and files

Reviewer: ```suggestion
   1. After the prepublish script is completed, unpack the release zip created from the release script (i.e. `cmsgov-design-system-2.0.0.tgz`) and check all relevant folders and files are present.
```

>       - cmsgov-design-system includes `dist`, `src` folders and `package.json`, `README.md` files

Reviewer: I know Scott suggested this, but I personally dont think its necessary to try and specify the files/folders, because it honestly depends. There could be additional files that are added in the future that arent accounted for here (i.e. the `types` or `locales` folders in HC.gov child ds), and the files/folders listed here are always subject to change. To a certain extent, the person releasing the design system has to know what is needed in the release

Reviewer: However, one thing that might be important to note is double checking that the release zips for `design-system` contain the `dist` folder

Author: I was thinking about the cases when we have changes, however, I think having this top level files would be helpful during the release. How about wording it as it typically would include, that way the content can change:
`The packages would typically contain:`
      - cmsgov-design-system: `dist`, `src` folders and `package.json`, `README.md` files
      - cmsgov-design-system-docs: `src` folder and `package.json`, `README.md` files
      - cmsgov-design-system-scripts: `gulp`, `jest` folders and `cli.js`, `configDefaults.js`, `package.json`, `README.md` files
`

Reviewer: I personally dont think this extra information is helpful. The important stuff to check is inside `dist`, and again that depends on the design system. 

`src`, `package.json` and `README.md` will always be there, and the files in `design-system-scripts` aren't manipulated by the build scripts at all. Again, the only important files to check are inside `dist`.

All of these files are subject to change meaning this documentation is going to be quickly outdated. Considering how it makes the release process document longer and more verbose, I would prefer to remove it.

      - cmsgov-design-system-docs includes `src` folder and `package.json`, `README.md` files
      - cmsgov-design-system-scrits includes `gulp`, `jest` folders and `cli.js`, `configDefaults.js`, `package.json`, `README.md` files

   1. The next step is to run the lerna publish command to publish to NPM.

      ```
      yarn release
      ```

>    1. After the publish process is completed, check [NPM](https://www.npmjs.com/package/@cmsgov/design-system) to verify the package is updated.

Reviewer: I dont think this is necessary, as the publish script will tell you if its failed or not

>    1. Next, [create a PR](https://github.com/CMSgov/design-system/blob/master/CONTRIBUTING.md#submitting-a-pull-request) for the release branch (i.e. `release-2.0.0`).
> 
>       - Add a title (keep the same name as the release branch, i.e. `Release-2.0.0`) and short description for your changes.
> 
>    1. Make sure to merge the PR (i.e. `Release-2.0.0`) into master after Reviewer approval process.

Reviewer: I think this could be a bit more concise, how does this sound?

```suggestion
   1. Create a PR for the release branch titled after the release (i.e. `Release 2.0.0`). Merge the PR into master after the changes are approved.
```

      ```
      This script will install dependencies, build design system assets, and run the linter and tests. If everything passes, `lerna` will prompt you for the new version number, which should follow the [SemVer release format](#versioning). After the versions are updated, the script will create a tagged release commit (prefaced with `core-`), a branch (prefaced with `release-`) containing the release commit, and a zip for the Github release notes.

>    1. The last step in the script will prompt you to continue with publishing to NPM. Proceed if everything is ready.

Reviewer: I wonder if its less confusing to remove the `yarn release` command, and just break it into `yarn prepublish` and `yarn publish`. This way we can include this step: "visually check that release zip created from the release script (i.e. `cmsgov-design-system-2.0.0.tgz`) contains the folders and files" before publishing files

Author: That's a good suggestion, it'll help add another level of checking in breaking `yarn release` into `yarn prepublish` and `yarn publish`

Author: @bernardwang found that `prepublish` and `publish` are NPM life cycle events, so what do you think if we change the process to `yarn pre-release` and `yarn release`?

Reviewer: Can we just use the life cycle events? So something like this:
```
    "prepublish": "./prepublish.sh",
    "publish": "yarn lerna publish from-git",
```
Then going forward in order to publish the release you can just run `yarn run publish` or `npm run publish`, which will run prepublish and publish lifecycle scripts.

Reviewer: Actually, I guess if we use the lifecycle events it wont be possible to separate the 2 commands because the publish lifecycle script automatically runs the prepublish lifecycel. Still, I kind of like the simplicity of using the lifecycle events, and you can always  directly run the commands (i.e. run `yarn lerna publish from-git` instead of `npm run publish`)

Author: That would be great, but when I used `"prepublish": "./prepublish.sh"` , running `yarn install` ended up executing the prepublish script

Author: Looks like `yarn install` calls the `prepublish`, I'm thinking it's okay to stay with `pre-release` & `release` for now and consider the `prepublish` hook for another PR?

Reviewer: @nichia I forgot to mention this, we should be using `prepublishOnly` instead of `prepublish`
Heres more background: https://docs.npmjs.com/misc/scripts

Author: I've changed package.json to have the `prepulishOnly` and `publish` script, however, getting an error message when I run `yarn publish`
```yarn publish
yarn publish v1.22.5
error Package marked as private, not publishing.```

Found this issue [issue about it](https://github.com/yarnpkg/yarn/issues/5334) but I cannot find where we specify `private:true` on the package.json.

Reviewer: Sorry again for the back and forth, but I saw this comment and we (shouldnt be overriding the lifecycle scripts)[https://github.com/lerna/lerna/issues/1969#issuecomment-498694766]. Maybe its best to keep it to what you had, which is `prerelease` and `release`. Also, please also update the script to be called `prerelease.sh` as well
